proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,332 @@
+      mtc.prompt_for_task("question-generation", some_unknown_param=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ValueError                                Traceback (most recent call last)
+      <ipython-input-16-369cca52e960> in <module>
+            1 # tasks parameters are checked
+      ----> 2 sa(some_param=[Document("Berlin is the capital of Germany.")])
+
+      2 frames
+      /usr/local/lib/python3.8/dist-packages/haystack/nodes/llm/multi_task.py in __call__(self, *args, **kwargs)
+           34         if set(template_dict.keys()) != set(self.prompt_params):
+           35             available_params = set(list(template_dict.keys()) + list(set(kwargs.keys())))
+      ---> 36             raise ValueError(f"Expected prompt params {self.prompt_params} but got {list(available_params)}")
+           37
+           38         template_dict["prompt_template"] = self.prompt_text
+
+      ValueError: Expected prompt params ['documents'] but got ['some_unknown_param']
+  ```
+
+- MultiTaskPipelines
+
+  Even though we can use MultiTaskComponent directly its real power lies in using pipelines and Haystack. For example, we
+  can retrieve documents from the document store using the query and then inject the retrieved documents into documents
+  as a parameter to the selected MultiTaskComponent task. For example:
+
+  ``` python
+      from haystack.pipelines import MultiTaskPipeline
+      top_k = 3
+      query = "Who are the parents of Arya Stark?"
+      mtc = MultiTaskComponent(current_task="question-answering")
+      pipe = MultiTaskPipeline(multi_tasks=[mtc], retriever)
+      output = pipe.run(query=query,
+                        params={"Retriever": {"top_k": top_k}},
+                        questions=[query for n in range(0, top_k)],
+                        #documents parameter we need for this task will be automatically populated by the retriever
+                        )
+
+      output["results"]
+  ```
+
+  However, we are still not utilizing the full power of Haystack pipelines. What if we could use more than
+  MultiTaskComponent in the pipeline? Perhaps we could first retrieve documents from the retriever, pass it
+  to first MultiTaskComponent component that will generate questions from these documents and then add a
+  second MultiTaskComponent component that will answer those generated questions given the documents as the
+  context. Here is how we can do exactly that:
+
+  ``` python
+      top_k = 3
+      query = "Who are the parents of Arya Stark?"
+      qg = MultiTaskComponent(current_task="question-generation")
+      qa = MultiTaskComponent(current_task="question-answering")
+
+      pipe = Pipeline()
+      pipe.add_node(component=retriever, name="Retriever", inputs=["Query"])
+      pipe.add_node(component=qg, name="qg", inputs=["Retriever"])
+      pipe.add_node(component=qa, name="qa", inputs=["qg"])
```

Comment:
Just to clarify here, do we envision that the `question-answering` node would receive both the generated questions and the originally retrieved documents to be able to answer those questions? If so we will need to think how to influence the output of MultiTaskComponent to optionally return its input. Or we could pass in the Retriever as input to the `question-answering` node as well.

Reply:
question answering node will have its template parameters injected either from retriever or from nodes prior to it in the pipeline. Is that what you are asking?